pkg/controller/storagecluster/update.go

```diff
@@ -393,31 +395,67 @@ func (c *Controller) getUnavailableNumbers(
 			"update of storage cluster  %#v: %v", cluster, err)
 	}
 
+	storageNodeList, err := c.Driver.GetStorageNodes(cluster)
+	if err != nil {
+		return -1, -1, fmt.Errorf("couldn't get list of storage nodes during rolling "+
+			"update of storage cluster  %#v: %v", cluster, err)
+	}
+	storageNodeMap := make(map[string]*storageapi.StorageNode)
+	for _, storageNode := range storageNodeList {
+		storageNodeMap[storageNode.SchedulerNodeName] = storageNode
+	}
+
 	var numUnavailable, desiredNumberScheduled int
 	for _, node := range nodeList.Items {
+		// If the node has a storage node and it is not healthy.
+		storageNode, storageNodeExists := storageNodeMap[node.Name]
+		if storageNodeExists {
+			delete(storageNodeMap, node.Name)
+			if storageNode.Status != storageapi.Status_STATUS_OK {
+				logrus.WithField("StorageNode", storageNode).Info("Storage node is not healthy")
+				numUnavailable++
+				continue
+			}
+		}
+
 		wantToRun, _, err := c.nodeShouldRunStoragePod(&node, cluster)
+		logrus.WithFields(logrus.Fields{
+			"node":              node.Name,
+			"wantToRun":         wantToRun,
+			"storageNodeExists": storageNodeExists,
+		}).WithError(err).Debug("check node should run storage pod")
 		if err != nil {
 			return -1, -1, err
 		}
-		if !wantToRun {
-			continue
-		}
-		desiredNumberScheduled++
-		storagePods, exists := nodeToStoragePods[node.Name]
-		if !exists {
-			numUnavailable++
-			continue
-		}
-		available := false
-		for _, pod := range storagePods {
-			// for the purposes of update we ensure that the pod is
-			// both available and not terminating
-			if podutil.IsPodReady(pod) && pod.DeletionTimestamp == nil {
-				available = true
-				break
+
+		// During Openshift cluster upgrade, the node will be tainted hence wantToRun will be false,
+		// so we need to check if StorageNode exists to decide whether to check if the storage node is available.
+		if wantToRun || storageNodeExists {
+			desiredNumberScheduled++
+			storagePods, exists := nodeToStoragePods[node.Name]
+			if !exists {
+				numUnavailable++
+				continue
+			}
+			available := false
+			for _, pod := range storagePods {
+				// for the purposes of update we ensure that the pod is
+				// both available and not terminating
+				if podutil.IsPodReady(pod) && pod.DeletionTimestamp == nil {
+					available = true
+					break
+				}
+			}
+			if !available {
+				numUnavailable++
 			}
 		}
-		if !available {
+	}
+
+	// For the storage nodes that do not have a corresponding k8s node.
+	for _, storageNode := range storageNodeMap {
+		if storageNode.Status != storageapi.Status_STATUS_OK {
```

Comment:
Should we increment the `desiredNumberScheduled` too? Or you are thinking of keeping it conservative and calculate the maxUnavailable number based only on kubernetes nodes and not portworx nodes?
I think that's fine too. Just thinking out loud if ever numUnavailable will always exceed maxUnavailable resulting in no way for users to override and force upgrade other online nodes. I mean they won't be able to go over 100%. This may happen if few non-k8s portworx nodes are offline for a known reason and users still want to continue with an upgrade for rest of the nodes. Having said that, they can always user numbers instead of percentage in that case and let the upgrade continue. So not a big deal, just wanted to know your reason too.

Reply:
Yes, in this case the external nodes will be upgraded externally and to be conservative admin can finish and ensure that is successful first.